atomic_reactor/plugins/post_push_to_pulp.py

```diff
@@ -100,6 +100,50 @@ def __init__(self, tasker, workflow, pulp_registry_name=None, load_squashed_imag
 
         self.pulp_handler = get_pulp_session(self.workflow, self.log, self.pulp_fallback)
 
+    def _deduplicate_layers(self, layers, filename, file_extension):
+        # getImageIdsExist was introduced in rh-dockpulp 0.6+
+        existing_imageids = self.pulp_handler.get_image_ids_existing(layers)
+        self.log.debug("existing layers: %s", existing_imageids)
+
+        # Strip existing layers from the tar and repack it
+        remove_layers = [str(os.path.join(x, 'layer.tar')) for x in existing_imageids]
+
+        commands = {'.xz': 'xzcat', '.gz': 'zcat', '.bz2': 'bzcat', '.tar': 'cat'}
+        unpacker = commands.get(file_extension, None)
+        self.log.debug("using unpacker %s for extension %s", unpacker, file_extension)
+        if unpacker is None:
+            raise Exception("Unknown tarball format: %s" % filename)
+
+        fd, upload_filename = tempfile.mkstemp(prefix='strip_tar_', suffix='.gz')
+        os.close(fd)
+        cmd = "set -o pipefail; {0} {1} | tar --delete {2} | gzip - > {3}".format(
+            unpacker, filename, ' '.join(remove_layers), upload_filename)
+        self.log.debug("running %s", cmd)
+        try:
+            subprocess.check_call(cmd, shell=True)
+            return upload_filename
+        except Exception as exc:
+            self._unlink_file(upload_filename)
+            raise(exc)
+
+    def _gzip_file(self, filename, file_extension):
+        if file_extension == '.tar':
+            fd, upload_filename = tempfile.mkstemp(prefix='full_tar_', suffix='.gz')
+            os.close(fd)
+            cmd = "set -o pipefail; cat {0} | gzip - > {1}".format(
+                filename, upload_filename)
+            self.log.debug("Compressing tar using '%s' command", cmd)
+            try:
+                subprocess.check_call(cmd, shell=True)
+                return upload_filename
+            except Exception as exc:
+                self._unlink_file(upload_filename)
+                raise(exc)
+
+    def _unlink_file(self, filename):
+        if filename and os.path.exists(filename):
+            os.unlink(filename)
```

Comment:
Instead of checking the file exists, how about just catching exceptions?
```
if filename:
    try:
        os.unlink(filename)
    except OSError:
        pass
```

Reply:
sure